src/LinuxTracing/LibunwindstackMaps.cpp

```diff
@@ -18,14 +20,78 @@ class LibunwindstackMapsImpl : public LibunwindstackMaps {
   unwindstack::Maps* Get() override { return maps_.get(); }
 
   void AddAndSort(uint64_t start, uint64_t end, uint64_t offset, uint64_t flags,
-                  const std::string& name, uint64_t load_bias) override {
-    maps_->Add(start, end, offset, flags, name, load_bias);
-    maps_->Sort();
-  }
+                  const std::string& name) override;
 
  private:
   std::unique_ptr<unwindstack::BufferMaps> maps_;
 };
+
+void LibunwindstackMapsImpl::AddAndSort(uint64_t start, uint64_t end, uint64_t offset,
+                                        uint64_t flags, const std::string& name) {
+  // First, remove existing maps that are fully contained in the new map, and resize or split
+  // existing maps that intersect with the new map. This is how the kernel handles memory mappings
+  // when using `mmap` with `MAP_FIXED` causes a new map that overlaps with existing ones. From the
+  // manpage of `mmap`: "`MAP_FIXED` [...] If the memory region specified by `addr` and `length`
+  // overlaps pages of any existing mapping(s), then the overlapped part of the existing mapping(s)
+  // will be discarded."
+  for (auto map_info_it = maps_->begin(); map_info_it != maps_->end();) {
+    std::shared_ptr<unwindstack::MapInfo> map_info = *map_info_it;
+    if (end <= map_info->start() || start >= map_info->end()) {
+      // The new map does not intersect map_info. Keep map_info untouched (do nothing).
+      ++map_info_it;
+
```

Comment:
Nit: Remove the empty line here and below (or add it in the else case)

Reply:
Hehe, you are not the only one to not like blank lines between `if`-`else if` cases but I think they help readability and the style guide agrees ("Blank lines inside a chain of if-else blocks may well help readability."). Adding it before the last brace of the `else` doesn't make a lot of sense though as it's the last case.

Reply:
I am actually fine with line-breaks there. I was wondering about the two lines in the scopes in the else case. At least in the first scope.

Reply:
Which ones?